packages/wallet/src/services/WalletAssetProvider.ts

```diff
@@ -0,0 +1,214 @@
+
+  // TODO: We need to fetch assets one by one because the provider will throw if any of the assets requests to the getAssets endpoint is not found.
+  // We want to fetch the ones we can and return a simplified AssetInfo for the ones we can't.
+  for (const assetId of assetIds) {
+    try {
+      const fetchedAsset = await provider.getAsset({
+        assetId,
+        extraData: { nftMetadata: true, tokenMetadata: true }
+      });
+      assetsFromProvider.push(fetchedAsset);
+    } catch (error) {
+      logger.error(error);
+    }
+  }
+
+  return assetsFromProvider;
+};
+
+const createFallbackAsset = (assetId: Cardano.AssetId): Asset.AssetInfo => {
+  const name = Cardano.AssetId.getAssetName(assetId);
+  const policyId = Cardano.AssetId.getPolicyId(assetId);
+  return {
+    assetId,
+    fingerprint: Cardano.AssetFingerprint.fromParts(policyId, name),
+    name,
+    policyId,
+    quantity: 0n,
+    supply: 0n
+  };
+};
+
+const mergeAssets = (
+  assetIds: Cardano.AssetId[],
+  cachedAssetsInfo: Map<Cardano.AssetId, Asset.AssetInfo>,
+  assetsFromProvider: Asset.AssetInfo[],
+  mintedAssets: Asset.AssetInfo[] | null
+): Asset.AssetInfo[] =>
+  assetIds.map((assetId) => {
+    const asset = cachedAssetsInfo.get(assetId) || assetsFromProvider.find((a) => a.assetId === assetId);
+    const mintedAsset = mintedAssets?.find((info) => info.assetId === assetId);
+
+    if (!asset && !mintedAsset) {
+      return createFallbackAsset(assetId);
+    }
+
+    if (!asset && mintedAsset) {
+      return mintedAsset;
+    }
+
+    if (asset && mintedAsset) {
+      asset.supply += mintedAsset.supply;
+
+      if (mintedAsset.nftMetadata) {
+        asset.nftMetadata = mintedAsset.nftMetadata;
+      }
+    }
+
+    return asset!;
+  });
+
+export const createWalletAssetProvider = ({
+  assetProvider,
+  assetInfo$,
+  tx,
+  logger
+}: AssetProviderContext): AssetProvider => ({
+  async getAsset({ assetId }: GetAssetArgs): Promise<Asset.AssetInfo> {
+    const mintedAssets = await getMintedAssetInfosFromTx(tx, logger);
+    const cachedAssetsInfo = await firstValueFrom(assetInfo$);
+
+    let asset = cachedAssetsInfo.get(assetId);
+
+    if (!asset) {
+      try {
+        asset = await assetProvider.getAsset({ assetId, extraData: { nftMetadata: true, tokenMetadata: true } });
+      } catch (error) {
+        logger.error(error);
+      }
+    }
+
+    const mintedAsset = mintedAssets?.find((info) => info.assetId === assetId);
+
+    // Let's create dummy AssetInfo for the unresolved asset. This is probably better than throwing as the UI can still present it as regular token.
+    if (!asset && !mintedAsset) {
+      const name = Cardano.AssetId.getAssetName(assetId);
+      const policyId = Cardano.AssetId.getPolicyId(assetId);
+      return {
+        assetId,
+        fingerprint: Cardano.AssetFingerprint.fromParts(policyId, name),
+        name,
+        policyId,
+        quantity: 0n,
+        supply: 0n
```

Comment:
How about this? Same suggestion for createFallbackAsset. Either way, I think it's important to add a doc comment explaining this behavior when asset info couldn't be resolved, because without looking at the code I would expect it to throw instead.
```suggestion
        quantity: Number.NaN,
        supply: Number.NaN
```

Reply:
I cant assing Number.NaN since those are bigints, but I think 0 supply could be also good? since 0 supply is not valid (tokens must have at least 1 to exists on chain). I did add the documentation as suggested tho:

https://github.com/input-output-hk/cardano-js-sdk/pull/1075/commits/c39bcf81694122eb4d49fef57ab00e41b3480634